### PR TITLE
UI-7645 - Refactor calculated measures migration

### PR DESCRIPTION
**Pull request by antoinetissier**
Before: the logic to migrate calculated measures is defined in the self-sufficient `migrateCalculatedMeasures`, in which a lot of the logic to go through and update each widget/dashboard is duplicated.
After: the logic to migrate calculated measures is split in two parts => 1. update the widget/dashboard MDX within the corresponding callbacks to avoid duplicating that logic, and 2. migrate the `calculated_measures` folder itself.

One of the nice things about not duplicating the logic of going through each dashboard/widget, is that it avoids having to re-parse/stringify and serialize/deserialize each saved dashboard and widget.
It also avoids the immutability issues described [here](https://github.com/activeviam/activeui-migration/pull/65#issuecomment-1362984014).
Since this logic is not tested at the calculated measures level anymore, a bunch of test files are removed.
Ultimately, in a separate PR, we should test the whole `migrate_50_to_51` migration.

### CODE REVIEW

File: src/5.0_to_5.1/__test_resources__/contentServer.ts [hunk 1]
```diff
@@ -299,56 +299,6 @@ export const contentServer: ContentRecord = {
                     },
                   },
                 },
-                "2sp": {
```
Comment (antoinetissier) [inline src/5.0_to_5.1/__test_resources__/contentServer.ts:302]: This is a folder with a single calculated measure in it.
But there was no entry in `ui/calculated_measures/content` matching this calculated measure (id `5m9`).
So I am simply removing it.

File: src/5.0_to_5.1/migrate_5.0_to_5.1.ts [hunk 2]
```diff
@@ -1,12 +1,38 @@
...
+    doesReportIncludeStacks,
+  },
 ) => {
-  // TODO migrate calculated measures
+  const namesOfCalculatedMeasuresToMigrate =
+    getNamesOfCalculatedMeasuresToMigrate(contentServer);
+  // Accumulate the cubes to which the saved calculated measures belong.
+  const measureToCubeMapping: { [measureName: string]: string[] } = {};
+
   migrateDashboards(migrateDashboard);
```
Comment (antoinetissier) [inline src/5.0_to_5.1/migrate_5.0_to_5.1.ts:22]: Ultimately we'll need to wire the variables related to calculated measures here as well.
But I prefer to do it in the separate PR where we make the whole thing work (and factorizing the mutation logic between `migrateWidget` and `migrateDashboard`).

**Review by eliasinho — COMMENTED** [review]
Looks great! Nice job on removing the code duplications and including the calculated measures in the generic migrate widgets/dashboards processes!

Have you tried to run the CLI and see if this works in its current state (for what's implemented)?

File: src/5.0_to_5.1/calculated-measures/migrateSavedCalculatedMeasures.test.ts [hunk 3]
```diff
@@ -1,25 +1,16 @@
...
-const dataModels = {
-  "Ranch 6.0": sandboxDataModel,
-  "Ranch 5.11": sandboxDataModel,
-};
 const contentServerForTests = _cloneDeep(contentServer);
 const errorReport: ErrorReport = {};
 
 contentServerForTests.children!.ui.children = {
   ...contentServerForTests.children!.ui.children,
   calculated_measures: uiCalculatedMeasuresFolder,
```
[the remark this answers is not in the document]
Comment (antoinetissier) [reply src/5.0_to_5.1/calculated-measures/migrateSavedCalculatedMeasures.test.ts:13]: Ultimately we could, and same for the `auiDashboardsFolder` that was there before (instead of removing it).
But this is even more refactor, and I prefer to keep this for later.

File: src/5.0_to_5.1/calculated-measures/getNamesOfCalculatedMeasuresToMigrate.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ */
+export function getNamesOfCalculatedMeasuresToMigrate(
+  contentServer: ContentRecord,
+): string[] {
+  const calculatedMeasuresFolder =
+    contentServer.children?.ui?.children?.calculated_measures;
+
+  const { content, structure } = calculatedMeasuresFolder?.children ?? {};
+  if (!content?.children || !structure?.children) {
+    return [];
```
[the remark this answers is not in the document]
Comment (antoinetissier) [reply src/5.0_to_5.1/calculated-measures/getNamesOfCalculatedMeasuresToMigrate.ts:19]: What can happen is that one specific dashboard is corrupt: i.e. it has an entry in `dashboards/content` but no corresponding entry in  `dashboards/structure`.
But having a `dashboards/content` folder existing without a `dashboards/structure` folder at all as a whole will never happen in real life.

File: src/5.0_to_5.1/calculated-measures/getNamesOfCalculatedMeasuresToMigrate.test.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import { contentServer } from "../__test_resources__/contentServer";
+import { getNamesOfCalculatedMeasuresToMigrate } from "./getNamesOfCalculatedMeasuresToMigrate";
+
+describe("getNamesOfCalculatedMeasuresToMigrate", () => {
+  it("returns the names corresponding to the calculated measure files created with ActiveUI 5.0", () => {
```
[the remark this answers is not in the document]
Comment (antoinetissier) [reply src/5.0_to_5.1/calculated-measures/getNamesOfCalculatedMeasuresToMigrate.test.ts:5]: I wanted to concey that it's only the saved ones, as opposed to query-scoped ones.
I'll do that by just adding "saved" with your wording, instead of my heavier one 👍

File: src/5.0_to_5.1/calculated-measures/getNamesOfCalculatedMeasuresToMigrate.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  }
+
+  const structureRecords = findContentRecords(
+    structure,
+    Object.keys(content.children),
+  );
+
+  return _map(structureRecords, ({ pathToParentFolder }, id) => {
+    const metadata = _getMetaData(structure, pathToParentFolder, id);
```
[the remark this answers is not in the document]
Comment (antoinetissier) [reply src/5.0_to_5.1/calculated-measures/getNamesOfCalculatedMeasuresToMigrate.ts:28]: `getMetaData` expects the node's content to be parsed, but it's not the case here so I am using `_getMetaData` instead.
It makes it consistent with the rest of the codebase.

File: src/5.0_to_5.1/calculated-measures/migrateCalculatedMeasuresInMdx.ts [hunk 7]
```diff
@@ -20,20 +20,22 @@ export const migrateCalculatedMeasuresInMdx = ({
...
-  namesOfCalculatedMeasurestoMigrate: string[];
+
+  namesOfCalculatedMeasuresToMigrate: string[];
 }): {
   cubeName: CubeName;
-  namesOfCalculatedMeasuresToMigrateInWidget: string[];
+  namesOfCalculatedMeasuresRemovedFromMdx: string[];
   migratedMdx: MdxSelect | MdxDrillthrough;
 } => {
   const namesOfCalculatedMeasuresToMigrateInWidget = _intersection(
```
[the remark this answers is not in the document]
Comment (antoinetissier) [reply src/5.0_to_5.1/calculated-measures/migrateCalculatedMeasuresInMdx.ts:36]: > Actually, we could even rename migrateCalculatedMeasuresInMdx in removeCalculatedMeasuresFromMdx

The arguments for `migrateCalculatedMeasuresInMdx ` were:
- `removeCalculatedMeasuresFromMdx` would be misleading: we're only removing the definitions of some of the CMs from the WHERE clause, but they are still used on the axes
-  have a consistent naming between`migrateCalculatedMeasuresInWidget` and `migrateCalculatedMeasuresInMdx`
- convey that we're also interested of gathering which calculated measure definitions were effectively removed, and what cube they were targeting

File: src/5.0_to_5.1/calculated-measures/migrateCalculatedMeasuresInMdx.test.ts [hunk 8]
```diff
@@ -96,11 +96,12 @@ describe("migrateCalculatedMeasuresInMdx", () => {
   it("returns `namesOfCalculatedMeasuresToMigrateInWidget` as an empty array and does not modify the MDX when there are no calculated measures", () => {
     const {
       cubeName,
-      namesOfCalculatedMeasuresToMigrateInWidget,
+      namesOfCalculatedMeasuresRemovedFromMdx:
+        namesOfCalculatedMeasuresToMigrateInWidget,
```
[the remark this answers is not in the document]
Comment (antoinetissier) [reply src/5.0_to_5.1/calculated-measures/migrateCalculatedMeasuresInMdx.test.ts:100]: I agree, this was lost in translation during the refactor.

File: src/5.0_to_5.1/calculated-measures/migrateCalculatedMeasuresInMdx.test.ts [hunk 9]
```diff
@@ -2,7 +2,7 @@ import { sandboxDataModel } from "@activeviam/data-model-5.1/dist/__test_resourc
 import { MdxSelect, parse, stringify } from "@activeviam/mdx-5.0";
 import { migrateCalculatedMeasuresInMdx } from "./migrateCalculatedMeasuresInMdx";
 
-const mdxSelectWithNoCalculatedMeasures: MdxSelect = parse(`
+export const mdxSelectWithNoCalculatedMeasures: MdxSelect = parse(`
```
[the remark this answers is not in the document]
Comment (antoinetissier) [reply src/5.0_to_5.1/calculated-measures/migrateCalculatedMeasuresInMdx.test.ts:5]: It's only used in two files that are next to each other, and it's small utils. I'd prefer to leave it for now and make it available in `__test_resources__` later if needed.
This way we avoid having some MDX defined locally in the `migrateCalculatedMeasuresInMdx` test and some MDX defined globally, or all MDX defined globally but part of it being used only in one file.

**Comment by antoinetissier** [comment]
> Have you tried to run the CLI and see if this works in its current state (for what's implemented)?

I intend to do that when everything is wired, in [this branch](https://github.com/activeviam/activeui-migration/tree/personal/a.tissier/UI-7645-final-wiring).

**Comment by antoinetissier** [comment]
> What do you mean by "the variables related to calculated measures"?

`measureToCubeMapping` and `namesOfCalculatedMeasuresToMigrate`

File: src/5.0_to_5.1/calculated-measures/migrateSavedCalculatedMeasures.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (eliasinho) [reply src/5.0_to_5.1/calculated-measures/migrateSavedCalculatedMeasures.test.ts:13]: Sounds fair!

File: src/5.0_to_5.1/calculated-measures/getNamesOfCalculatedMeasuresToMigrate.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (eliasinho) [reply src/5.0_to_5.1/calculated-measures/getNamesOfCalculatedMeasuresToMigrate.ts:19]: > What can happen is that one specific dashboard is corrupt: i.e. it has an entry in dashboards/content but no corresponding entry in dashboards/structure.

Ah right, that was the use case I had in mind, but it's not reflected here, my bad

File: src/5.0_to_5.1/calculated-measures/migrateCalculatedMeasuresInMdx.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (eliasinho) [reply src/5.0_to_5.1/calculated-measures/migrateCalculatedMeasuresInMdx.ts]: Hmm, okay to keep `migrateCalculatedMeasuresInMdx`, but then I'd change `namesOfCalculatedMeasuresRemovedFromMdx` to be consistent. What about `namesOfMigratedCalulatedMeasures`?

File: src/5.0_to_5.1/calculated-measures/migrateCalculatedMeasuresInMdx.ts [hunk 10]
```diff
@@ -20,28 +20,28 @@ export const migrateCalculatedMeasuresInMdx = ({
...
 }: {
   mdx: MdxSelect | MdxDrillthrough;
   serverKey?: string;
   dataModels: { [serverKey: string]: DataModel };
-  namesOfCalculatedMeasurestoMigrate: string[];
+  namesOfCalculatedMeasuresToMigrate: string[];
 }): {
   cubeName: CubeName;
-  namesOfCalculatedMeasuresToMigrateInWidget: string[];
+  namesOfCalculatedMeasuresRemovedFromMdx: string[];
```
Comment (eliasinho) [inline src/5.0_to_5.1/calculated-measures/migrateCalculatedMeasuresInMdx.ts:31]: Answering [this comment](https://github.com/activeviam/activeui-migration/pull/70#discussion_r1061759629).

I'm happy to keep `migrateCalculatedMeasuresInMdx`, but then I'd change `namesOfCalculatedMeasuresRemovedFromMdx` to be consistent. What about `namesOfMigratedCalulatedMeasures`?

File: src/5.0_to_5.1/calculated-measures/migrateCalculatedMeasuresInMdx.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (eliasinho) [reply src/5.0_to_5.1/calculated-measures/migrateCalculatedMeasuresInMdx.test.ts:5]: Okay to keep them here for now

File: src/5.0_to_5.1/calculated-measures/getNamesOfCalculatedMeasuresToMigrate.ts [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+    // TODO https://support.activeviam.com/jira/browse/UI-8110 Remove the cast when `name` is mandatory in the SDK.
```
[the remark this answers is not in the document]
Comment (antoinetissier) [reply src/5.0_to_5.1/calculated-measures/getNamesOfCalculatedMeasuresToMigrate.ts:29]: TODO with link to a ticket is a convention that we have.
It strikes your eye and allows to quickly find all occurrences with a global search, so I would like to keep it.
I can mention that this should be changed in `activeui-migration` in the ticket though.

File: src/5.0_to_5.1/calculated-measures/getNamesOfCalculatedMeasuresToMigrate.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (antoinetissier) [reply src/5.0_to_5.1/calculated-measures/getNamesOfCalculatedMeasuresToMigrate.ts:28]: It's insignificant. And I prefer to do the parsing on a per file basis, only when needed, through `_getMetaData` rather than doing it once and for all on the whole content server and then using `getMetaData` like we do in ActiveUI. It's a different use case: it's more performant in ActiveUI, but it would be less performant here.

File: src/5.0_to_5.1/calculated-measures/migrateCalculatedMeasuresInMdx.ts (hunk 10 above)
Comment (antoinetissier) [reply src/5.0_to_5.1/calculated-measures/migrateCalculatedMeasuresInMdx.ts:31]: Done ✅

File: src/5.0_to_5.1/calculated-measures/getNamesOfCalculatedMeasuresToMigrate.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (eliasinho) [reply src/5.0_to_5.1/calculated-measures/getNamesOfCalculatedMeasuresToMigrate.ts:29]: Okay, as long as this is mentioned in the ticket, I'm fine with leaving the comment like this.
It's more that I don't believe that relying **only** on a "TODO" comment for a task to be done is efficient. But if there's always a link to a ticket and that the ticket is comprehensive, you're right, it's probably even better actually.

**Review by eliasinho — APPROVED** [review]
Looks good to me! 🚀🚀🚀
